### PR TITLE
fix(manager/poetry): support fixed version constraint

### LINKED ISSUE #19972: Setting a Poetry constraint causes lockfile update to fail
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, tell us what version of Renovate you run.

Present on self-hosted and github app

### If you're self-hosting Renovate, select which platform you are using.

gitlab.com

### If you're self-hosting Renovate, tell us what version of the platform you run.

Also observed on gitlab.com 15.8

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

When setting a python and poetry constraint, the poetry lock update command fails because renovate attempts to pip install poetry using the wrong format for pip version constraints. See the repro here: https://github.com/WarpRat/renovate-poetry-contraint-repro/pull/6

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: Failed to update poetry.lock file(branch="renovate/pylint-2.x")
{
  "err": {
    "name": "ExecError",
    "cmd": "/bin/sh -c docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/mnt/renovate/gh/WarpRat/renovate-poetry-contraint-repro\":\"/mnt/renovate/gh/WarpRat/renovate-poetry-contraint-repro\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -v \"/tmp/containerbase\":\"/tmp/containerbase\" -e PIP_CACHE_DIR -e BUILDPACK_CACHE_DIR -e CONTAINERBASE_CACHE_DIR -w \"/mnt/renovate/gh/WarpRat/renovate-poetry-contraint-repro\" docker.io/renovate/sidecar bash -l -c \"install-tool python 3.9.9 && pip install --user poetry1.3.2 && poetry update --lock --no-interaction pylint\"",
    "stderr": "ERROR: Could not find a version that satisfies the requirement poetry1.3.2 (from versions: none)\nERROR: No matching distribution found for poetry1.3.2\n",
    "stdout": "installing v2 tool python v3.9.9\nlinking tool python v3.9.9\nPython 3.9.9\npip 22.3.1 from /opt/buildpack/tools/python/3.9.9/lib/python3.9/site-packages/pip (python 3.9)\nInstalled v2 /usr/local/buildpack/tools/v2/python.sh in 24 seconds\nskip cleanup, not a docker build: 8d08c5ae1170\n",
    "options": {
      "cwd": "/mnt/renovate/gh/WarpRat/renovate-poetry-contraint-repro",
      "encoding": "utf-8",
      "env": {
        "PIP_CACHE_DIR": "/tmp/renovate-cache/others/pip",
        "HOME": "/home/ubuntu",
        "PATH": "/home/ubuntu/.local/bin:/home/ubuntu/bin:/opt/buildpack/tools/python/3.9.3/bin:/home/ubuntu/.npm-global/bin:/home/ubuntu/renovateapp/node_modules/.bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
        "LC_ALL": "C.UTF-8",
        "LANG": "C.UTF-8",
        "BUILDPACK_CACHE_DIR": "/tmp/containerbase",
        "CONTAINERBASE_CACHE_DIR": "/tmp/containerbase"
      },
      "maxBuffer": 10485760,
      "timeout": 900000
    },
    "exitCode": 1,
    "message": "Command failed: docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/mnt/renovate/gh/WarpRat/renovate-poetry-contraint-repro\":\"/mnt/renovate/gh/WarpRat/renovate-poetry-contraint-repro\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -v \"/tmp/containerbase\":\"/tmp/containerbase\" -e PIP_CACHE_DIR -e BUILDPACK_CACHE_DIR -e CONTAINERBASE_CACHE_DIR -w \"/mnt/renovate/gh/WarpRat/renovate-poetry-contraint-repro\" docker.io/renovate/sidecar bash -l -c \"install-tool python 3.9.9 && pip install --user poetry1.3.2 && poetry update --lock --no-interaction pylint\"\nERROR: Could not find a version that satisfies the requirement poetry1.3.2 (from versions: none)\nERROR: No matching distribution found for poetry1.3.2\n",
    "stack": "ExecError: Command failed: docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/mnt/renovate/gh/WarpRat/renovate-poetry-contraint-repro\":\"/mnt/renovate/gh/WarpRat/renovate-poetry-contraint-repro\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -v \"/tmp/containerbase\":\"/tmp/containerbase\" -e PIP_CACHE_DIR -e BUILDPACK_CACHE_DIR -e CONTAINERBASE_CACHE_DIR -w \"/mnt/renovate/gh/WarpRat/renovate-poetry-contraint-repro\" docker.io/renovate/sidecar bash -l -c \"install-tool python 3.9.9 && pip install --user poetry1.3.2 && poetry update --lock --no-interaction pylint\"\nERROR: Could not find a version that satisfies the requirement poetry1.3.2 (from versions: none)\nERROR: No matching distribution found for poetry1.3.2\n\n    at ChildProcess.<anonymous> (/home/ubuntu/renovateapp/node_modules/renovate/dist/util/exec/common.js:87:24)\n    at ChildProcess.emit (node:events:525:35)\n    at ChildProcess.emit (node:domain:489:12)\n    at Process.ChildProcess._handle.onexit (node:internal/child_process:293:12)"
  }
}
```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction repository in the bug description

### PR DESCRIPTION
**Pull request by rarkins**
## Changes

Massages fixed version poetry constraints

## Context

Closes #19972

Replaces #19973

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/poetry/artifacts.ts [hunk 1]
```diff
@@ -187,7 +192,7 @@ export async function updateArtifacts({
       docker: {},
       toolConstraints: [{ toolName: 'python', constraint }],
       preCommands: [
-        `pip install --user ${quote(`poetry${poetryVersion ?? ''}`)}`,
+        `pip install --user ${quote(`poetry${poetryConstraint ?? ''}`)}`,
```
Comment (viceice) [inline lib/modules/manager/poetry/artifacts.ts:195]: can we convert this to tool install instead? containerbase is prepared.

- https://github.com/renovatebot/renovate/pull/19973

Comment (rarkins) [reply lib/modules/manager/poetry/artifacts.ts:195]: Solve the same problem?

Comment (viceice) [reply lib/modules/manager/poetry/artifacts.ts:195]: yes

Comment (rarkins) [reply lib/modules/manager/poetry/artifacts.ts]: ok, rewritten

**Review by viceice — APPROVED** [review]

**Pull request merged**